### ISSUE TITLE
Logstash1.5 - Removing tag in filter fails with undefined method delete

### ISSUE THREAD

**Issue by thehybridtechnician** [elastic/logstash#2261]
Upgraded to 1.5Beta1 to test out functionality.  Attempting to remove a tag fails.  Below is a snippet from my configuration.

Filter configuration
filter {
  if [program] == "services_monitor" and [message] =~ /Current memory used by supervisor/ {
    grok {
      patterns_dir            => "/etc/logstash/patterns/dellPatterns"
      match           => [
        "message", "Current memory used by supervisor %{DATA}\[%{DATA:Process}\] = %{NUMBER:memUsedKB:int}"
      ]
      remove_tag      => ["logs"]
      add_tag         => ["state"]
    }
  }
}

Failure Message
Exception in filterworker {"exception"=>#<NoMethodError: undefined method `delete' for #<Java::JavaUtil::ArrayList:0x6ada79a9>>, "backtrace"=>["/opt/logstash/lib/logstash/filters/base.rb:198:in `filter_matched'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/lib/logstash/filters/base.rb:193:in `filter_matched'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-grok-0.1.2/lib/logstash/filters/grok.rb:287:in `filter'", "(eval):187175:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):187171:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):187168:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):187129:in `initialize'", "org/jruby/RubyProc.java:271:in `call'", "/opt/logstash/lib/logstash/pipeline.rb:268:in `filter'", "/opt/logstash/lib/logstash/pipeline.rb:206:in `filterworker'", "/opt/logstash/lib/logstash/pipeline.rb:150:in `start_filters'"], :level=>:error}

Thank you!

**Comment by talevy** [comment]
hey @thehybridtechnician! Just checking-in to see if this is still an issue.

**Comment by suyograo** [comment]
I tried to reproduce this too, but could not. Please re-open this if this is still an issue

**Comment by ruddj** [comment]
Just wanted to add I experienced the same error in 1.5b1 using mutate and remove_tag option
`      mutate { 
        replace => ["type", "procurve" ] 
        remove_tag => [ "procurve" ]  
      }`
Exception is:
```
{:timestamp=>"2015-02-05T11:04:51.742000+1100", 
:message=>"Exception in filterworker", 
"exception"=>#<NoMethodError: undefined method `delete' for #<Java::JavaUtil::ArrayList:0x571e66ba>>, 
"backtrace"=>["/opt/logstash/lib/logstash/filters/base.rb:198:in `filter_matched'", 
"org/jruby/RubyArray.java:1613:in `each'", 
"/opt/logstash/lib/logstash/filters/base.rb:193:in `filter_matched'",
"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-mutate-0.1.2/lib/logstash/filters/mutate.rb:220:in `filter'", 
"(eval):30754:in `initialize'", 
"org/jruby/RubyArray.java:1613:in `each'", 
"org/jruby/RubyEnumerable.java:805:in `flat_map'", 
"(eval):30750:in `initialize'", 
"org/jruby/RubyArray.java:1613:in `each'", 
"org/jruby/RubyEnumerable.java:805:in `flat_map'", 
"(eval):30747:in `initialize'", 
"org/jruby/RubyArray.java:1613:in `each'", 
"org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):30707:in `initialize'", 
"org/jruby/RubyProc.java:271:in `call'", 
"/opt/logstash/lib/logstash/pipeline.rb:268:in `filter'", 
"/opt/logstash/lib/logstash/pipeline.rb:206:in `filterworker'", 
"/opt/logstash/lib/logstash/pipeline.rb:150:in `start_filters'"], 
:level=>:error}
```

**Comment by suyograo** [comment]
@ruddj thanks for the feedback. can you please provide a sample event which reproduces with this config?

**Comment by ruddj** [comment]
The problem is that is a small part of one config file of several files and custom patterns.

Here is that full file for recognising Procurve switch entries:
```ruby
filter {
 ### Match Switches ###
  if [type] == "syslog" {
    if !("procurve" in [tags]) {
      grok {
        match => [
          "host", "%{IP:src_ip}"
        ] 
        tag_on_failure => []
      }
      if [src_ip] {
        cidr {
          add_tag => [ "procurve" ]
          address => [ "%{src_ip}" ]
          network => [ "192.168.10.224/27" ]
        }
        mutate { remove_field => [ "src_ip" ]}
      }
    }

    ####### Procurve Switches ####
    if "procurve" in [tags] {
      # Change type to Procurve and strip out excess tags
      mutate { 
        replace => ["type", "procurve" ] 
        remove_tag => [ "procurve" ]  # seems to crash Logstash 1.5b1
      }
      grok {
        match => [
          "syslog_program", "(?:%{PROCURVE_EVENT:procurve_event})%{PROCURVE_TAG:procurve_tag}", 
          "syslog_program", "%{PROCURVE_TAG:procurve_tag}"
        ] 
        tag_on_failure => ["_grokparsefailure","procurve_match_eventtag"]
        patterns_dir => ["/etc/logstash/patterns"]
      }

      # Extract common fields from each message
      grok {
        break_on_match => false
        tag_on_failure => []
        match => [
          "syslog_message", "(trunk|port) %{PROCURVE_PORT:procurve_port}",
          "syslog_message", "%{PROCURVE_MAC:procurve_src_mac}",
          "syslog_message", "(vlan|vid) %{INT:procurve_vlan:int}"
        ] 
        patterns_dir => ["/etc/logstash/patterns"]
      }

      # Match with more complex expressions
      grok {
        match => [
          "syslog_message", "%{PROCURVE_PORT_CYCLE}",
          "syslog_message", "%{PROCURVE_DIPLD_00981}",
          "syslog_message", "%{PROCURVE_AUTH}",
          "syslog_message", "%{PROCURVE_LLDP_00828}",
          "syslog_message", "%{PROCURVE_ARPPRT_00913}",
          "syslog_message", "%{PROCURVE_ARPPRT_00917}",
          "syslog_message", "%{PROCURVE_ACL}",
          "syslog_message", "%{PROCURVE_IP_DROP}"
        ] 
        patterns_dir => ["/etc/logstash/patterns"]
        #add_tag => [ "procurve_match" ]
        tag_on_failure => []
      }

      if [procurve_port] {
        mutate { add_field => { "procurve_global_port_id" => "%{host}-%{procurve_port}" } }
      }
    }
    ###### End of Procurve Switches #######
  }
}
```
Here are the custom ProCurve patterns for it
```ruby
# hp procurve switch log files
PROCURVE %{TIMESTAMP_ISO8601:syslog_timestamp} <%{WORD:syslog_facility}.%{WORD:syslog_severity}> %{SYSLOGHOST:syslog_hostname} %{PROG:syslog_program}:  %{GREEDYDATA:syslog_message}

PROCURVE_TAGGED_SYSLOG ^<%{POSINT:syslog_pri:int}> %{CISCOTIMESTAMP:syslog_timestamp} %{SYSLOGHOST:syslog_hostname} (?:%{PROCURVE_EVENT:procurve_event})%{PROCURVE_TAG:procurve_tag}:
PROCURVE_TAGGED_SYSLOG_DEBUG ^<%{POSINT:syslog_pri:int}> %{CISCOTIMESTAMP:syslog_timestamp} %{SYSLOGHOST:syslog_hostname} %{PROCURVE_TAG:procurve_tag}(?:[ ]*):
PROCURVE_TAG (?:[-\w]+)
PROCURVE_EVENT [0-9]{5} 
PROCURVE_PORT Trk(?:[1-9][0-9]*)|(?:[0-9A-La-l]+)
PROCURVE_MAC (?:[A-Fa-f0-9]{6}-[A-Fa-f0-9]{6})

# Individual Filters
PROCURVE_DIPLD_00981 Access denied %{IP:procurve_src_ip} -> %{IP:procurve_dst_ip}
PROCURVE_ARPPRT_00913 dest:%{PROCURVE_MAC:procurve_dst_mac} target:%{PROCURVE_MAC:procurve_tgt_mac}
PROCURVE_ARPPRT_00917 invalid IP %{IP:procurve_src_ip}
PROCURVE_AUTH_TYPE login|logout
PROCURVE_AUTH User '%{USERNAME:user}' %{PROCURVE_AUTH_TYPE:procurve_action}(( from %{IP:procurve_src_ip})|$)
PROCURVE_LLDP_00828 port %{PROCURVE_PORT:procurve_local_port}\(VID %{POSINT:procurve_local_vlan:int}\)with peer device port %{PROCURVE_PORT:procurve_peer_port}\(VID %{POSINT:procurve_peer_vlan:int}\)
PROCURVE_IP_ACT fwd|drop
PROCURVE_IP_DROP s=%{IP:procurve_src_ip}\(vl%{INT:procurve_src_vlan:int}\) d=%{IP:procurve_dst_ip}\(vl%{INT:procurve_dst_vlan:int}\) (?:g=%{IP:procurve_gtw_ip} )?(?:%{PROCURVE_IP_ACT:procurve_action})?(?:[ ]*)(?:%{GREEDYDATA:procurve_drp_reason})?
PROCURVE_PORT_STATE on-line|off-line|(Blocked by STP)
PROCURVE_PORT_CYCLE port %{PROCURVE_PORT} is (?:now )?%{PROCURVE_PORT_STATE:procurve_port_state}
PROCURVE_ACL_ACTION permitted|denied
PROCURVE_ACL_PROT ip|udp|tcp|icmp|igmp
PROCURVE_ACL List %{USER:procurve_acl_rule}, seq#%{INT:procurve_acl_rulenum:int} %{PROCURVE_ACL_ACTION:procurve_acl_action} %{PROCURVE_ACL_PROT:procurve_acl_protocol} %{IP:procurve_src_ip}(?:\(%{INT:procurve_src_port:int}\))? -> %{IP:procurve_dst_ip}(?:\(%{INT:procurve_dst_port:int}\))?
```

Some sample input:
```
<134> Feb  5 15:32:34 192.168.10.249 00076 ports:  port A13 is now on-line
<132> Feb  5 15:32:28 192.168.10.249 02672 FFI:  port A13-Excessive link state transitions
<134> Feb  5 15:26:14 192.168.10.247 arp-protect:  Deny ARP Req 4437e6-b4efcb,169.254.114.32 port B11, vlan 50
<134> Feb  5 15:21:08 192.168.10.247 arp-protect:  Deny pkt, invalid IP 0.0.0.0 port B1, vlan 50
<132> Feb  5 14:20:24 192.168.10.236 00981 dipld:  Access denied 169.254.20.184 -> 224.0.0.22 port 26, 1 packets received since last log.
```

**Comment by jmherbst** [comment]
I'm running into this same issue while running logstash 1.5.0.rc2.  Could this please be re-opened?  

Here is the my testing logstash config:
```
input {
  stdin {
    codec => json { charset => "CP1252" }
  }
}

filter {
  if [message] =~ /test/ {
    mutate {
      add_tag => ["TEST"]
      remove_tag => ["tag2"]
    }
  }
}

output {
  stdout { codec => rubydebug }
}
```

Here is my `testing.log` that I'm piping into logstash via `stdin`:
```{"source_ip": "127.0.0.1","source": "testing","program": "PROG","priority": "notice","message": "This is just a test message","host_from": "ls-ship-testing","host": "test.example.com","facility": "user","date": "Mar 30 05:00:03.874","@version": "1","@timestamp": "2015-04-02T14:37:59.708Z","tags":["tag1","tag2"]}```

Here is the output of logstash with `--debug` flag:
```
Writing garbage collection logs to ./logstash-gc.log
Adding plugin path {:path=>"/home/jherbst/logstash-1.5.0.rc2", :level=>:debug, :file=>"logstash/agent.rb", :line=>"291", :method=>"configure_plugin_path"}
Reading config file {:file=>"logstash/lib/logstash/agent.rb", :level=>:debug, :line=>"334", :method=>"local_config"}
Compiled pipeline code:
        @inputs = []
        @filters = []
        @outputs = []
        @periodic_flushers = []
        @shutdown_flushers = []

          @input_stdin_1 = plugin("input", "stdin", LogStash::Util.hash_merge_many({ "codec" => plugin("codec", "json", LogStash::Util.hash_merge_many({ "charset" => ("CP1252".force_encoding(Encoding::UTF_8)) })) }))

          @inputs << @input_stdin_1

          @filter_mutate_2 = plugin("filter", "mutate", LogStash::Util.hash_merge_many({ "add_tag" => [("TEST".force_encoding(Encoding::UTF_8))] }, { "remove_tag" => [("tag2".force_encoding(Encoding::UTF_8))] }))

          @filters << @filter_mutate_2

            @filter_mutate_2_flush = lambda do |options, &block|
              @logger.debug? && @logger.debug("Flushing", :plugin => @filter_mutate_2)

              flushed_events = @filter_mutate_2.flush(options)

              return if flushed_events.nil? || flushed_events.empty?

              flushed_events.each do |event|
                @logger.debug? && @logger.debug("Flushing", :plugin => @filter_mutate_2, :event => event)

                events = [event]


                block.call(event)
                events.flatten.each{|e| block.call(e) if e != event}
              end

            end

            if @filter_mutate_2.respond_to?(:flush)
              @periodic_flushers << @filter_mutate_2_flush if @filter_mutate_2.periodic_flush
              @shutdown_flushers << @filter_mutate_2_flush
            end

          @output_stdout_3 = plugin("output", "stdout", LogStash::Util.hash_merge_many({ "codec" => ("rubydebug".force_encoding(Encoding::UTF_8)) }))

          @outputs << @output_stdout_3

  @filter_func = lambda do |event, &block|
    events = [event]
    @logger.debug? && @logger.debug("filter received", :event => event.to_hash)
            events = events.flat_map do |event|
              events = [event]
              if (((event["[message]"] =~ Regexp.new(("test".force_encoding(Encoding::UTF_8)))))) # if [message] =~ /test/
                events = events.flat_map do |event|
                  next [] if event.cancelled?

                  new_events = []
                  @filter_mutate_2.filter(event){|new_event| new_events << new_event}
                  event.cancelled? ? new_events : new_events.unshift(event)
                end


              end
              events
            end

    events.flatten.each{|e| block.call(e) }
  end
  @output_func = lambda do |event, &block|
    events = [event]
    @logger.debug? && @logger.debug("output received", :event => event.to_hash)
    @output_stdout_3.handle(event)

  end {:level=>:debug, :file=>"logstash/lib/logstash/pipeline.rb", :line=>"30", :method=>"initialize"}
Plugin not defined in namespace, checking for plugin file {:type=>"codec", :name=>"json", :path=>"logstash/codecs/json", :level=>:debug, :file=>"logstash/lib/logstash/plugin.rb", :line=>"133", :method=>"lookup"}
Using version 0.1.x codec plugin 'json'. This plugin isn't well supported by the community and likely has no maintainer. {:level=>:info, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"223", :method=>"print_version_notice"}
config LogStash::Codecs::JSON/@charset = "CP1252" {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
Plugin not defined in namespace, checking for plugin file {:type=>"input", :name=>"stdin", :path=>"logstash/inputs/stdin", :level=>:debug, :file=>"logstash/lib/logstash/plugin.rb", :line=>"133", :method=>"lookup"}
Using version 0.1.x input plugin 'stdin'. This plugin isn't well supported by the community and likely has no maintainer. {:level=>:info, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"223", :method=>"print_version_notice"}
config LogStash::Inputs::Stdin/@codec = <LogStash::Codecs::JSON charset=>"CP1252"> {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Inputs::Stdin/@debug = false {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Inputs::Stdin/@add_field = {} {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
Plugin not defined in namespace, checking for plugin file {:type=>"filter", :name=>"mutate", :path=>"logstash/filters/mutate", :level=>:debug, :file=>"logstash/lib/logstash/plugin.rb", :line=>"133", :method=>"lookup"}
Using version 0.1.x filter plugin 'mutate'. This plugin isn't well supported by the community and likely has no maintainer. {:level=>:info, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"223", :method=>"print_version_notice"}
config LogStash::Filters::Mutate/@add_tag = ["TEST"] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@remove_tag = ["tag2"] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@type = "" {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@tags = [] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@exclude_tags = [] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@add_field = {} {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@remove_field = [] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Filters::Mutate/@periodic_flush = false {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
Plugin not defined in namespace, checking for plugin file {:type=>"output", :name=>"stdout", :path=>"logstash/outputs/stdout", :level=>:debug, :file=>"logstash/lib/logstash/plugin.rb", :line=>"133", :method=>"lookup"}
Using version 0.1.x output plugin 'stdout'. This plugin isn't well supported by the community and likely has no maintainer. {:level=>:info, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"223", :method=>"print_version_notice"}
Plugin not defined in namespace, checking for plugin file {:type=>"codec", :name=>"rubydebug", :path=>"logstash/codecs/rubydebug", :level=>:debug, :file=>"logstash/lib/logstash/plugin.rb", :line=>"133", :method=>"lookup"}
Using version 0.1.x codec plugin 'rubydebug'. This plugin isn't well supported by the community and likely has no maintainer. {:level=>:info, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"223", :method=>"print_version_notice"}
config LogStash::Codecs::RubyDebug/@metadata = false {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Outputs::Stdout/@codec = <LogStash::Codecs::RubyDebug > {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Outputs::Stdout/@type = "" {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Outputs::Stdout/@tags = [] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Outputs::Stdout/@exclude_tags = [] {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Outputs::Stdout/@workers = 1 {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
Automatically switching from json to json_lines codec {:plugin=>"stdin", :level=>:info, :file=>"logstash/lib/logstash/inputs/base.rb", :line=>"133", :method=>"fix_streaming_codecs"}
Using version 0.1.x codec plugin 'json_lines'. This plugin isn't well supported by the community and likely has no maintainer. {:level=>:info, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"223", :method=>"print_version_notice"}
config LogStash::Codecs::JSONLines/@charset = "CP1252" {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
config LogStash::Codecs::Line/@charset = "UTF-8" {:level=>:debug, :file=>"logstash/lib/logstash/config/mixin.rb", :line=>"112", :method=>"config_init"}
Pipeline started {:level=>:info, :file=>"logstash/lib/logstash/pipeline.rb", :line=>"83", :method=>"run"}
Logstash startup completed
filter received {:event=>{"source_ip"=>"127.0.0.1", "source"=>"testing", "program"=>"PROG", "priority"=>"notice", "message"=>"This is just a test message", "host_from"=>"ls-ship-testing", "host"=>"test.example.com", "facility"=>"user", "date"=>"Mar 30 05:00:03.874", "@version"=>"1", "@timestamp"=>"2015-04-02T14:37:59.708Z", "tags"=>#<Java::JavaUtil::ArrayList:0x7cab2145>}, :level=>:debug, :file=>"(eval)", :line=>"45", :method=>"initialize"}
Plugin is finished {:plugin=><LogStash::Inputs::Stdin >, :level=>:info, :file=>"logstash/lib/logstash/plugin.rb", :line=>"61", :method=>"finished"}
filters/LogStash::Filters::Mutate: adding tag {:tag=>"TEST", :level=>:debug, :file=>"logstash/lib/logstash/filters/base.rb", :line=>"187", :method=>"filter_matched"}
filters/LogStash::Filters::Mutate: removing tag {:tag=>"tag2", :level=>:debug, :file=>"logstash/lib/logstash/filters/base.rb", :line=>"195", :method=>"filter_matched"}
stdin shutting down. {:level=>:debug, :file=>"logstash/vendor/bundle/jruby/1.9/gems/logstash-input-stdin-0.1.3/lib/logstash/inputs/stdin.rb", :line=>"42", :method=>"teardown"}
Exception in filterworker {"exception"=>#<NoMethodError: undefined method `delete' for #<Java::JavaUtil::ArrayList:0x7cab2145>>, "backtrace"=>["/home/jherbst/logstash-1.5.0.rc2/logstash/lib/logstash/filters/base.rb:197:in `filter_matched'", "org/jruby/RubyArray.java:1613:in `each'", "/home/jherbst/logstash-1.5.0.rc2/logstash/lib/logstash/filters/base.rb:192:in `filter_matched'", "/home/jherbst/logstash-1.5.0.rc2/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-mutate-0.1.5/lib/logstash/filters/mutate.rb:218:in `filter'", "(eval):53:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):49:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):46:in `initialize'", "org/jruby/RubyProc.java:271:in `call'", "/home/jherbst/logstash-1.5.0.rc2/logstash/lib/logstash/pipeline.rb:275:in `filter'", "/home/jherbst/logstash-1.5.0.rc2/logstash/lib/logstash/pipeline.rb:209:in `filterworker'", "/home/jherbst/logstash-1.5.0.rc2/logstash/lib/logstash/pipeline.rb:153:in `start_filters'"], :level=>:error, :file=>"logstash/lib/logstash/pipeline.rb", :line=>"222", :method=>"filterworker"}
Plugin is finished {:plugin=><LogStash::Outputs::Stdout >, :level=>:info, :file=>"logstash/lib/logstash/plugin.rb", :line=>"61", :method=>"finished"}
Pipeline shutdown complete. {:level=>:info, :file=>"logstash/lib/logstash/pipeline.rb", :line=>"97", :method=>"run"}
Logstash shutdown completed
[root@ls2-index-testing jherbst]# cat testing.log
{"source_ip": "127.0.0.1","source": "testing","program": "PROG","priority": "notice","message": "This is just a test message","host_from": "ls-ship-testing","host": "test.example.com","facility": "user","date": "Mar 30 05:00:03.874","@version": "1","@timestamp": "2015-04-02T14:37:59.708Z","tags":["tag1","tag2"]}
```

**Comment by johntdyer** [comment]
:+1:

**Comment by suyograo** [comment]
was able to reproduce

```
suyog@machine:~/ws/logstash (master)$ ruby -v
jruby 1.7.16 (1.9.3p392) 2014-09-25 575b395 on Java HotSpot(TM) 64-Bit Server VM 1.7.0_60-b19 +jit [darwin-x86_64]
suyog@machine:~/ws/logstash (master)$ bin/logstash irb
1.9 :001 > require "logstash/json"
 => true
1.9 :002 > data = '{"foo": ["test1", "test2"]}'
 => "{\"foo\": [\"test1\", \"test2\"]}"
1.9 :003 > json_data = LogStash::Json.load(data)
 => {"foo"=>#<Java::JavaUtil::ArrayList:0x6bc80765>}
1.9 :004 > json_data["foo"] << "test3"
 => #<Java::JavaUtil::ArrayList:0x6bc80765>
1.9 :008 > json_data["foo"].to_s
 => "[test1, test2, test3]"
1.9 :009 > json_data["foo"].delete("test2")
NoMethodError: undefined method `delete' for #<Java::JavaUtil::ArrayList:0x6bc80765>
  from (irb):9:in `evaluate'
  from org/jruby/RubyKernel.java:1107:in `eval'
  from org/jruby/RubyKernel.java:1507:in `loop'
  from org/jruby/RubyKernel.java:1270:in `catch'
  from org/jruby/RubyKernel.java:1270:in `catch'
  from /Users/suyog/ws/logstash/lib/logstash/runner.rb:75:in `run'
  from org/jruby/RubyProc.java:271:in `call'
  from /Users/suyog/ws/logstash/lib/logstash/runner.rb:129:in `run'
  from org/jruby/RubyProc.java:271:in `call'
  from /Users/suyog/ws/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.19/lib/stud/task.rb:12:in `initialize'
1.9 :006 > json_data["foo"].remove("test2")
 => true
1.9 :007 > json_data["foo"].to_s
 => "[test1]"
```

**Comment by colinsurprenant** [comment]
this is related to the new JrJackson json serialization when the input data is json. the created objects are Java objects, like ArrayList, and it does not support the delete method. we have to check where will be the best place to add support for delete with the Ruby semantic. I think we could add it here https://github.com/jruby/jruby/blob/master/core/src/main/ruby/jruby/java/java_ext/java.util.rb and submit upstream and add it localling in the interim.

**Comment by colinsurprenant** [comment]
fixed by #2991

**Comment by inptensib** [comment]
Hi everybody ! I am trying to insert a json in mongoDb using Logstash-output-mongodb.Mt json objects contains arrayLists fields do i'm getting thi exception:
<NoMethodError: undefined method `bson_type' for #<Java::JavaUtil::ArrayList:0x745ad733>>,
I am using logstash 1.5.1.Thank you !!!
